Databases/MongoDB - error 500 on creds generation when mongodb replicaset primary step-down

I have following environmnent setup:
3 node mongodb replicaset
3 node vault (v0.7.3) cluster with store on 3 server consul (0.8.3) cluster
Configured databases - mongodb secret backend.
My anonymised connection_url:
connection_url:mongodb://USER:PASS@x.x.x.x:27017,y.y.y.y:27017,z.z.z.z:27017/admin?replicaSet=REPLICASETNAME
Problem:
After configuration it works perfect. But when mongodb steps down primary and select new one I receive following error when I try generate credentials:

```
 Error reading mongodb/creds/payments: Error making API request.

URL: GET http://x.x.x.x:8200/v1/mongodb/creds/payments
Code: 500. Errors:

* 1 error occurred:

* EOF
```

I temporary fix situation by running vault step-down. After that vault starts generate credentials.

Logs when I try to release 2 instances of service (java spring with spring cloud vault configuration) which using databases/mongodb backend. In log I have only two entries in this step:

```
July 5th 2017, 09:14:11.572	2017/07/05 07:14:11.572405 [ERROR] sys: lease revocation failed: lease_id=XXX_mongo/creds/payments/bc91c10b-1c19-3a8f-0996-3c64abaa1d83 error=failed to revoke entry: resp:(*logical.Response)(nil) err:EOF
July 5th 2017, 09:14:13.996	2017/07/05 07:14:13.997640 [ERROR] sys: lease revocation failed: lease_id=XXX_mongo/creds/payments/4d1e45b1-55b1-7fe8-c6b4-31204721eb84 error=failed to revoke entry: resp:(*logical.Response)(nil) err:EOF
```

After few minutes I decide to try step-down current master:

```
July 5th 2017, 09:38:32.121	2017/07/05 07:38:32.122704 [WARN ] core: stepping down from active operation to standby
July 5th 2017, 09:38:32.127	2017/07/05 07:38:32.127418 [INFO ] core: pre-seal teardown starting
July 5th 2017, 09:38:32.128	2017/07/05 07:38:32.127535 [INFO ] core: forwarding rpc listeners stopped
July 5th 2017, 09:38:32.128	2017/07/05 07:38:32.127443 [INFO ] core: shutting down forwarding rpc listeners
July 5th 2017, 09:38:32.128	2017/07/05 07:38:32.127436 [INFO ] core: stopping cluster listeners
July 5th 2017, 09:38:32.417	2017/07/05 07:38:32.418210 [INFO ] core: rpc listeners successfully shut down
July 5th 2017, 09:38:32.417	2017/07/05 07:38:32.417797 [WARN ] transport: http2Client.notifyError got notified that the client transport was broken EOF.
July 5th 2017, 09:38:32.421	2017/07/05 07:38:32.419659 [WARN ] grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: Error while dialing dial tcp X.X.X.X:8201: getsockopt: connection refused"; Reconnecting to {X.X.X.X:8201 <nil>}
July 5th 2017, 09:38:32.421	2017/07/05 07:38:32.418242 [INFO ] core: cluster listeners successfully shut down
July 5th 2017, 09:38:32.421	2017/07/05 07:38:32.420441 [WARN ] transport: http2Client.notifyError got notified that the client transport was broken EOF.
July 5th 2017, 09:38:32.423	2017/07/05 07:38:32.421964 [WARN ] grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: Error while dialing dial tcp X.X.X.X:8201: getsockopt: connection refused"; Reconnecting to {X.X.X.X:8201 <nil>}
July 5th 2017, 09:38:32.424	2017/07/05 07:38:32.418552 [INFO ] core: pre-seal teardown complete
July 5th 2017, 09:38:32.424	2017/07/05 07:38:32.418284 [INFO ] rollback: stopping rollback manager
July 5th 2017, 09:38:32.433	2017/07/05 07:38:32.432937 [INFO ] core: acquired lock, enabling active operation
July 5th 2017, 09:38:32.436	2017/07/05 07:38:32.436680 [ERROR] core: failed to acquire lock: error=failed to acquire lock: Unexpected response code: 500 (rpc error: invalid session "b52232cb-1648-d3df-61d4-4d0224c9beb2")
July 5th 2017, 09:38:32.519	2017/07/05 07:38:32.519770 [INFO ] core: post-unseal setup starting
July 5th 2017, 09:38:32.520	2017/07/05 07:38:32.520157 [WARN ] grpc: addrConn.transportMonitor exits due to: context canceled
July 5th 2017, 09:38:32.522	2017/07/05 07:38:32.522130 [INFO ] core: loaded wrapping token key
July 5th 2017, 09:38:32.526	2017/07/05 07:38:32.526579 [INFO ] core: successfully mounted backend: type=generic path=secret/
July 5th 2017, 09:38:32.528	2017/07/05 07:38:32.526949 [INFO ] core: successfully mounted backend: type=system path=sys/
July 5th 2017, 09:38:32.529	2017/07/05 07:38:32.527140 [INFO ] core: successfully mounted backend: type=rabbitmq path=XXX_rabbitmq/
July 5th 2017, 09:38:32.529	2017/07/05 07:38:32.527046 [INFO ] core: successfully mounted backend: type=database path=XXX_mongo/
July 5th 2017, 09:38:32.532	2017/07/05 07:38:32.527201 [INFO ] core: successfully mounted backend: type=consul path=XXX_consul/
July 5th 2017, 09:38:32.546	2017/07/05 07:38:32.527520 [INFO ] rollback: starting rollback manager
July 5th 2017, 09:38:32.546	2017/07/05 07:38:32.527276 [INFO ] core: successfully mounted backend: type=cubbyhole path=cubbyhole/
July 5th 2017, 09:38:32.548	2017/07/05 07:38:32.540222 [INFO ] expiration: restoring leases
July 5th 2017, 09:38:32.571	2017/07/05 07:38:32.571062 [INFO ] expire: leases restored: restored_lease_count=34
July 5th 2017, 09:38:32.575	2017/07/05 07:38:32.575309 [INFO ] core: post-unseal setup complete
July 5th 2017, 09:38:32.576	2017/07/05 07:38:32.575591 [INFO ] core/startClusterListener: starting listener: listener_address=0.0.0.0:8201
July 5th 2017, 09:38:32.578	2017/07/05 07:38:32.575792 [INFO ] core/startClusterListener: serving cluster requests: cluster_listen_address=[::]:8201
July 5th 2017, 09:38:33.421	2017/07/05 07:38:33.422205 [WARN ] grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: Error while dialing dial tcp X.X.X.X:8201: getsockopt: connection refused"; Reconnecting to {X.X.X.X:8201 <nil>}
July 5th 2017, 09:38:33.855	2017/07/05 07:38:33.856121 [WARN ] grpc: addrConn.transportMonitor exits due to: context canceled
```

After that when I run `vault read XXX_mongo/creds/payments` from shell - vault generates new username and password.
Then I try to confirm that problem is in mongodb replicaset and I run rs.stepDown() on mongodb primary. After that I run same vault command with result (anonymised mount point):

```
Error reading XXX_mongo/creds/payments: Error making API request.

URL: GET http://X.X.X.X:8200/v1/XXX_mongo/creds/payments
Code: 500. Errors:

* 1 error occurred:

* EOF
```

But there is no results in logs for that operation. Next I step-down current vault master and it back to normal.
Maby vault after rs.stepDown() in mongo replicaset it try to connect to old primary and after remounting database backend it refresh which mongo node is primary and works valid?

--- a/plugins/database/mongodb/mongodb.go
+++ b/plugins/database/mongodb/mongodb.go
@@ -2,6 +2,7 @@ package mongodb
 
 import (
 	"io"
+	"strings"
 	"time"
 
 	"encoding/json"
@@ -128,7 +129,7 @@ func (m *MongoDB) CreateUser(statements dbplugin.Statements, usernameConfig dbpl
 	err = session.DB(mongoCS.DB).Run(createUserCmd, nil)
 	switch err {
 	case nil:
-	case io.EOF:
+	case io.EOF || strings.Contains(err.Error(), "EOF"):
 		if err := m.ConnectionProducer.Close(); err != nil {
 			return "", "", errwrap.Wrapf("error closing EOF'd mongo connection: {{err}}", err)
 		}
@@ -184,7 +185,7 @@ func (m *MongoDB) RevokeUser(statements dbplugin.Statements, username string) er
 	switch err {
 	case nil:
 	case mgo.ErrNotFound:
-	case io.EOF:
+	case io.EOF || strings.Contains(err.Error(), "EOF"):
 		if err := m.ConnectionProducer.Close(); err != nil {
 			return errwrap.Wrapf("error closing EOF'd mongo connection: {{err}}", err)
 		}

--- a/plugins/database/mongodb/mongodb.go
+++ b/plugins/database/mongodb/mongodb.go
@@ -1,12 +1,14 @@
 package mongodb
 
 import (
+	"io"
 	"time"
 
 	"encoding/json"
 
 	"fmt"
 
+	"github.com/hashicorp/errwrap"
 	"github.com/hashicorp/vault/api"
 	"github.com/hashicorp/vault/builtin/logical/database/dbplugin"
 	"github.com/hashicorp/vault/plugins"
@@ -124,7 +126,21 @@ func (m *MongoDB) CreateUser(statements dbplugin.Statements, usernameConfig dbpl
 	}
 
 	err = session.DB(mongoCS.DB).Run(createUserCmd, nil)
-	if err != nil {
+	switch err {
+	case nil:
+	case io.EOF:
+		if err := m.ConnectionProducer.Close(); err != nil {
+			return "", "", errwrap.Wrapf("error closing EOF'd mongo connection: {{err}}", err)
+		}
+		session, err := m.getConnection()
+		if err != nil {
+			return "", "", err
+		}
+		err = session.DB(mongoCS.DB).Run(createUserCmd, nil)
+		if err != nil {
+			return "", "", err
+		}
+	default:
 		return "", "", err
 	}
 
@@ -165,7 +181,22 @@ func (m *MongoDB) RevokeUser(statements dbplugin.Statements, username string) er
 	}
 
 	err = session.DB(db).RemoveUser(username)
-	if err != nil && err != mgo.ErrNotFound {
+	switch err {
+	case nil:
+	case mgo.ErrNotFound:
+	case io.EOF:
+		if err := m.ConnectionProducer.Close(); err != nil {
+			return errwrap.Wrapf("error closing EOF'd mongo connection: {{err}}", err)
+		}
+		session, err := m.getConnection()
+		if err != nil {
+			return err
+		}
+		err = session.DB(db).RemoveUser(username)
+		if err != nil {
+			return err
+		}
+	default:
 		return err
 	}
 

--- a/plugins/database/mongodb/mongodb.go
+++ b/plugins/database/mongodb/mongodb.go
@@ -127,9 +127,9 @@ func (m *MongoDB) CreateUser(statements dbplugin.Statements, usernameConfig dbpl
 	}
 
 	err = session.DB(mongoCS.DB).Run(createUserCmd, nil)
-	switch err {
-	case nil:
-	case io.EOF || strings.Contains(err.Error(), "EOF"):
+	switch {
+	case err == nil:
+	case err == io.EOF, strings.Contains(err.Error(), "EOF"):
 		if err := m.ConnectionProducer.Close(); err != nil {
 			return "", "", errwrap.Wrapf("error closing EOF'd mongo connection: {{err}}", err)
 		}
@@ -182,10 +182,9 @@ func (m *MongoDB) RevokeUser(statements dbplugin.Statements, username string) er
 	}
 
 	err = session.DB(db).RemoveUser(username)
-	switch err {
-	case nil:
-	case mgo.ErrNotFound:
-	case io.EOF || strings.Contains(err.Error(), "EOF"):
+	switch {
+	case err == nil, err == mgo.ErrNotFound:
+	case err == io.EOF, strings.Contains(err.Error(), "EOF"):
 		if err := m.ConnectionProducer.Close(); err != nil {
 			return errwrap.Wrapf("error closing EOF'd mongo connection: {{err}}", err)
 		}

